libs/math/include/math/base_types.hpp

```diff
@@ -242,5 +242,30 @@ static constexpr meta::IfIsFixedPoint<T, T> Type(std::string const &val)
   return T(val);
 }
 
+template <typename T, typename U>
+static constexpr meta::IfIsFloat<T, T> AsType(U val, meta::IfIsInteger<U> * /*unused*/ = nullptr)
+{
+  return static_cast<T>(val);
+}
+
+template <typename T, typename U>
+static constexpr meta::IfIsFixedPoint<T, T> AsType(U val,
+                                                   meta::IfIsInteger<U> * /*unused*/ = nullptr)
+{
+  return static_cast<T>(val);
+}
+
+template <typename T, typename U>
+static constexpr meta::IfIsFloat<T, T> AsType(U val, meta::IfIsFloat<U> * /*unused*/ = nullptr)
+{
+  return static_cast<T>(val);
+}
+
+template <typename T, typename U>
+static constexpr meta::IfIsFixedPoint<T, T> AsType(U val, meta::IfIsFloat<U> * /*unused*/ = nullptr)
+{
+  return T::FromFloat(val);
```

Comment:
I think we need to rework this slightly. We want AsType to be the ONLY way to type convert a fixed point to a float, and privatise T::FromFloat(123.123). This is because if we allow multiple ways of doing the type conversion we make it easier to miss one later on when we're trying to find all cases where a fixed point is constructed via type conversion from a float. It's also more explicit that this is a type conversion this way.

Therefore, I suggest you have a struct here (perhaps wrapped in an anonymous namespace) that is a friend of FixedPoint that does the T::FromFloat for you. Then all other calls to T::FromFloat should be calls to AsType<float/double>

Reply:
I see your point, I'll do the necessary changes.

Reply:
made FromFloat() private, now constructionfrom float is only available through AsType<>, fixed and pushed